File: plugins/papercuts/src/papercuts/store.py
from __future__ import annotations

import hashlib
import json
import os
import random
import socket
import time
import uuid
from contextlib import contextmanager
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Iterator, Sequence

from papercuts.models import PapercutsError

_KINDS = {"complaint", "encounter", "resolved", "reopened"}
_LOCK_TIMEOUT_SECONDS = 5.0
_STALE_LOCK_SECONDS = 300.0


class JournalStore:
    def __init__(self, path: Path) -> None:
        self.path = path
        self.lock_path = Path(f"{path}.lock")

    def read_events(self) -> list[dict[str, Any]]:
        """Parse complete UTF-8 JSON object lines and reject malformed interiors."""
        events, _ = self._read_events_and_tail()
        return events

    @contextmanager
    def mutation(self) -> Iterator[list[dict[str, Any]]]:
        """Acquire the adjacent lock directory, read current events, and release safely."""
        self._reject_symlinks()
        try:
            self.path.parent.mkdir(mode=0o700, parents=True, exist_ok=True)
        except OSError as error:
            raise _io_error("create journal directory", self.path.parent, error) from error

        token = self._acquire_lock()
        try:
            events, incomplete_tail = self._read_events_and_tail()
            if incomplete_tail:
                raise PapercutsError(
                    "malformed_journal",
                    f"Journal has an incomplete final record: {self.path}",
                    exit_status=65,
                    suggested_fix="Run doctor --repair-tail before mutating the journal.",
                )
            yield events
        finally:
            self._release_lock(token)

    def append_events_locked(self, events: Sequence[dict[str, Any]]) -> None:
        """Append complete compact JSON records, flush, and fsync before return."""
        if not events:
            return
        records = b"".join(
            (
                json.dumps(
                    event,
                    ensure_ascii=False,
                    separators=(",", ":"),
                    sort_keys=True,
                )
                + "\n"
            ).encode("utf-8")
            for event in events
        )
        try:
            descriptor = os.open(
                self.path,
                os.O_WRONLY | os.O_CREAT | os.O_APPEND,
                0o600,
            )
            with os.fdopen(descriptor, "ab") as journal:
                journal.write(records)
                journal.flush()
                os.fsync(journal.fileno())
        except (OSError, TypeError, ValueError) as error:
            raise _io_error("append journal", self.path, error) from error

    def digest(self) -> str:
        """Return SHA-256 of current journal bytes, with empty bytes for a missing file."""
        try:
            data = self.path.read_bytes() if self.path.exists() else b""
        except OSError as error:
            raise _io_error("read journal", self.path, error) from error
        return hashlib.sha256(data).hexdigest()

    def _read_events_and_tail(self) -> tuple[list[dict[str, Any]], bool]:
        try:
            data = self.path.read_bytes()
        except FileNotFoundError:
            return [], False
        except OSError as error:
            raise _io_error("read journal", self.path, error) from error

        lines = data.splitlines(keepends=True)
        incomplete_tail = bool(lines and not lines[-1].endswith(b"\n"))
        complete_lines = lines[:-1] if incomplete_tail else lines
        events: list[dict[str, Any]] = []
        for line_number, raw_line in enumerate(complete_lines, start=1):
            try:
                line = raw_line.removesuffix(b"\n").removesuffix(b"\r").decode(
                    "utf-8"
                )
                event = json.loads(line)
            except (UnicodeDecodeError, json.JSONDecodeError) as error:
                raise _malformed(self.path, line_number) from error
            _validate_event(event, self.path, line_number)
            events.append(event)
        return events, incomplete_tail

    def _acquire_lock(self) -> str:
        token = uuid.uuid4().hex
        deadline = time.monotonic() + _LOCK_TIMEOUT_SECONDS
        while True:
            try:
                self.lock_path.mkdir()
            except FileExistsError:
                self._recover_stale_lock()
                remaining = deadline - time.monotonic()
                if remaining <= 0:
                    raise PapercutsError(
                        "lock_timeout",
                        f"Timed out waiting for journal lock: {self.lock_path}",
                        exit_status=75,
                        retryable=True,
                    )
                time.sleep(min(remaining, random.uniform(0.025, 0.1)))
                continue
            except OSError as error:
                raise _io_error("create journal lock", self.lock_path, error) from error

            metadata = {
                "token": token,
                "pid": os.getpid(),
                "host": socket.gethostname(),
                "created_at": _utc_now(),
            }
            try:
                descriptor = os.open(
                    self._metadata_path(),
                    os.O_WRONLY | os.O_CREAT | os.O_EXCL,
                    0o600,
                )
                with os.fdopen(descriptor, "w", encoding="utf-8") as metadata_file:
                    metadata_file.write(
                        json.dumps(metadata, separators=(",", ":"), sort_keys=True)
                        + "\n"
                    )
            except OSError as error:
                try:
                    self.lock_path.rmdir()
                except OSError:
                    pass
                raise _io_error("write journal lock", self.lock_path, error) from error
            return token

    def _recover_stale_lock(self) -> None:
        metadata = self._read_lock_metadata()
        if metadata is None or metadata.get("host") != socket.gethostname():
            return
        try:
            created_at = datetime.fromisoformat(
                str(metadata["created_at"]).replace("Z", "+00:00")
            )
            age = datetime.now(timezone.utc) - created_at.astimezone(timezone.utc)
            pid = int(metadata["pid"])
            token = str(metadata["token"])
        except (KeyError, TypeError, ValueError):
            return
        if age.total_seconds() <= _STALE_LOCK_SECONDS or not _process_absent(pid):
            return
        self._remove_lock_if_owned(token)

    def _release_lock(self, token: str) -> None:
        self._remove_lock_if_owned(token)

    def _remove_lock_if_owned(self, token: str) -> None:
        metadata = self._read_lock_metadata()
        if metadata is None or metadata.get("token") != token:
            return
        try:
            self._metadata_path().unlink()
            self.lock_path.rmdir()
        except FileNotFoundError:
            return
        except OSError:
            return

    def _read_lock_metadata(self) -> dict[str, Any] | None:
        try:
            metadata = json.loads(self._metadata_path().read_text(encoding="utf-8"))
        except (OSError, UnicodeDecodeError, json.JSONDecodeError):
            return None
        return metadata if isinstance(metadata, dict) else None

    def _metadata_path(self) -> Path:
        return self.lock_path / "owner.json"

    def _reject_symlinks(self) -> None:
        target = self.path.absolute()
        candidates = [target]
        if target.parent != target:
            candidates.append(target.parent)
        while not os.path.lexists(candidates[-1]):
            parent = candidates[-1].parent
            if parent == candidates[-1]:
                break
            candidates.append(parent)
        for current in reversed(candidates):
            try:
                if current.is_symlink():
                    raise PapercutsError(
                        "invalid_input",
                        f"Journal path traverses a symbolic link: {current}",
                        exit_status=64,
                    )
            except OSError as error:
                raise _io_error("inspect journal path", current, error) from error


def fold_events(events: Sequence[dict[str, Any]]) -> dict[str, dict[str, Any]]:
    records: dict[str, dict[str, Any]] = {}
    for index, event in enumerate(events, start=1):
        _validate_event(event, Path("<events>"), index)
        kind = event["kind"]
        if kind == "complaint":
            complaint_id = event["id"]
            if complaint_id in records:
                raise _malformed(Path("<events>"), index, "duplicate complaint ID")
            records[complaint_id] = {
                **event,
                "status": "open",
                "encounter_count": 1,
                "vote_count": 0,
                "last_encounter_at": event["ts"],
                "resolution_note": None,
                "recent_encounters": [],
                "status_history": [],
            }
            continue

        complaint_id = event["complaint_id"]
        record = records.get(complaint_id)
        if record is None:
            raise _malformed(Path("<events>"), index, "orphan event")
        if event["project"]["id"] != record["project"]["id"]:
            raise _malformed(Path("<events>"), index, "mismatched project ID")

        if kind == "encounter":
            record["encounter_count"] += 1
            record["vote_count"] += 1
            record["last_encounter_at"] = event["ts"]
            record["recent_encounters"] = [
                *record["recent_encounters"],
                event,
            ][-10:]
        else:
            record["status"] = "resolved" if kind == "resolved" else "open"
            if kind == "resolved":
                record["resolution_note"] = event.get("note")
            record["status_history"].append(event)
    return records


def _validate_event(event: Any, path: Path, line_number: int) -> None:
    if not isinstance(event, dict):
        raise _malformed(path, line_number)
    project = event.get("project")
    if (
        event.get("contract") != 1
        or event.get("kind") not in _KINDS
        or not isinstance(event.get("ts"), str)
        or not event["ts"]
        or not isinstance(event.get("agent"), str)
        or not event["agent"]
        or not isinstance(project, dict)
        or not isinstance(project.get("id"), str)
        or not project["id"]
        or not isinstance(project.get("name"), str)
        or not project["name"]
    ):
        raise _malformed(path, line_number)

    kind = event["kind"]
    if kind == "complaint":
        valid = (
            isinstance(event.get("id"), str)
            and isinstance(event.get("text"), str)
            and event.get("severity") in {"minor", "major", "blocker"}
            and isinstance(event.get("tags"), list)
            and all(isinstance(tag, str) for tag in event["tags"])
            and isinstance(event.get("context"), dict)
        )
    else:
        valid = isinstance(event.get("complaint_id"), str)
    if not valid:
        raise _malformed(path, line_number)


def _process_absent(pid: int) -> bool:
    if pid <= 0:
        return False
    try:
        os.kill(pid, 0)
    except ProcessLookupError:
        return True
    except (PermissionError, OSError):
        return False
    return False


def _utc_now() -> str:
    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")


def _malformed(path: Path, line_number: int, detail: str | None = None) -> PapercutsError:
    suffix = f" ({detail})" if detail else ""
    return PapercutsError(
        "malformed_journal",
        f"Malformed journal event at {path}:{line_number}{suffix}",
        exit_status=65,
    )


def _io_error(action: str, path: Path, error: BaseException) -> PapercutsError:
    return PapercutsError(
        "io_failure",
        f"Could not {action} at {path}: {error}",
        exit_status=74,
        retryable=True,
    )

File: plugins/papercuts/src/store.js
import fs from 'node:fs';
import path from 'node:path';
import os from 'node:os';
import crypto from 'node:crypto';
import {hash, stableJSON, isObject, nonempty, invalid, PapercutsError} from './models.js';

const C = fs.constants;
const malformed = (file, line, detail = '') => new PapercutsError('malformed_journal', `Malformed journal event at ${file}:${line}${detail ? ` (${detail})` : ''}`);
const unsafe = file => invalid(`Storage must use private, user-owned real directories and regular files: ${file}`);
const uid = () => process.getuid?.();
function owned(stat) { return uid() === undefined || stat.uid === uid(); }
function privateDirectory(directory, create = false, home = false) {
  if (create) { try { fs.mkdirSync(directory, {mode: 0o700}); } catch (e) { if (e.code !== 'EEXIST') throw e; } }
  const stat = fs.lstatSync(directory);
  if (!stat.isDirectory() || !owned(stat) || (process.platform !== 'win32' && (stat.mode & (home ? 0o022 : 0o077)))) throw unsafe(directory);
  return stat;
}
function same(a,b) { return a.dev === b.dev && a.ino === b.ino; }
function regular(file, flags, {create = false, exclusive = false, missing = false} = {}) {
  let expected;
  try { expected = fs.lstatSync(file); } catch (e) { if (e.code !== 'ENOENT') throw e; }
  if (expected && (!expected.isFile() || !owned(expected) || expected.nlink !== 1 || (process.platform !== 'win32' && (expected.mode & 0o077)))) throw unsafe(file);
  let fd;
  try {
    const additions = (C.O_NOFOLLOW || 0) | (C.O_NONBLOCK || 0) | (create ? C.O_CREAT : 0) | (exclusive || (create && !expected) ? C.O_EXCL : 0);
    fd = fs.openSync(file, flags | additions, 0o600);
    const stat = fs.fstatSync(fd);
    if (!stat.isFile() || !owned(stat) || stat.nlink !== 1 || (expected && !same(stat, expected)) || (process.platform !== 'win32' && (stat.mode & 0o077))) throw unsafe(file);
    return fd;
  } catch (e) {
    if (fd !== undefined) fs.closeSync(fd);
    if (e.code === 'ENOENT' && missing) return null;
    if (['ELOOP','ENOTDIR'].includes(e.code)) throw unsafe(file);
    throw e;
  }
}
function syncDirectory(directory) {
  if (process.platform === 'win32') return;
  const fd = fs.openSync(directory, C.O_RDONLY | (C.O_DIRECTORY || 0) | (C.O_NOFOLLOW || 0));
  try { fs.fsyncSync(fd); } finally { fs.closeSync(fd); }
}
function writeFile(file, data) {
  const fd = regular(file, C.O_WRONLY, {create: true, exclusive: true});
  try { fs.writeFileSync(fd, data); fs.fsyncSync(fd); } finally { fs.closeSync(fd); }
}
export function validateEvent(event, file = '<events>', line = 1) {
  if (!isObject(event) || event.contract !== 1 || !['complaint','encounter','resolved','reopened'].includes(event.kind) || !nonempty(event.ts) || !nonempty(event.agent) || !isObject(event.project) || !nonempty(event.project.id) || !nonempty(event.project.name)) throw malformed(file,line);
  let valid;
  if (event.kind === 'complaint') valid = nonempty(event.id) && nonempty(event.text) && ['minor','major','blocker'].includes(event.severity) && Array.isArray(event.tags) && event.tags.length <= 10 && event.tags.every(nonempty) && isObject(event.context);
  else valid = nonempty(event.complaint_id) && (event.note == null || nonempty(event.note)) && (event.kind !== 'encounter' || isObject(event.context));
  if (!valid) throw malformed(file,line);
}
export function foldEvents(events, file = '<events>') {
  const records = new Map();
  events.forEach((event, index) => {
    validateEvent(event,file,index+1);
    if (event.kind === 'complaint') {
      if (records.has(event.id)) throw malformed(file,index+1,'duplicate complaint ID');
      records.set(event.id, {...event, status:'open', encounter_count:1, vote_count:0, last_encounter_at:event.ts, resolution_note:null, recent_encounters:[], status_history:[]});
    } else {
      const record = records.get(event.complaint_id);
      if (!record) throw malformed(file,index+1,'orphan event');
      if (record.project.id !== event.project.id) throw malformed(file,index+1,'mismatched project ID');
      if (event.kind === 'encounter') {
        record.encounter_count++; record.vote_count++; record.last_encounter_at = event.ts;
        record.recent_encounters = [...record.recent_encounters,event].slice(-10);
      } else {
        record.status = event.kind === 'resolved' ? 'resolved' : 'open';
        if (event.kind === 'resolved') record.resolution_note = event.note ?? null;
        record.status_history.push(event);
      }
    }
  });
  return records;
}
export function serializeEvents(events) {
  events.forEach((event,index) => validateEvent(event,'<events>',index+1));
  return Buffer.from(events.map(e => stableJSON(e) + '\n').join(''));
}
export class JournalStore {
  constructor(storage) {
    this.path = storage.journal_path; this.home = storage.home; this.directory = path.dirname(this.path); this.lockPath = this.path + '.lock';
    if (this.path !== path.join(this.home, `.${storage.client}`, 'papercuts.jsonl') || !['codex','claude'].includes(storage.client)) throw unsafe(this.path);
  }
  check(create = false) {
    privateDirectory(this.home, false, true);
    // ponytail: private user storage limits path races; use native directory-relative I/O if shared storage returns.
    const homeParent = path.dirname(this.home);
    let ancestor = homeParent;
    while (true) {
      const stat = fs.lstatSync(ancestor);
      if (!stat.isDirectory() || (process.platform !== 'win32' && (stat.mode & 0o022) && !(stat.mode & 0o1000))) throw unsafe(ancestor);
      const next = path.dirname(ancestor); if (next === ancestor) break; ancestor = next;
    }
    const stat = privateDirectory(this.directory, create, true);
    if (this.parentIdentity && !same(stat, this.parentIdentity)) throw unsafe(this.directory);
  }
  bytes() {
    try {
      this.check(); const fd = regular(this.path,C.O_RDONLY,{missing:true});
      if (fd === null) return Buffer.alloc(0);
      try { return fs.readFileSync(fd); } finally { fs.closeSync(fd); }
    } catch (e) { if (e.code === 'ENOENT') return Buffer.alloc(0); throw e; }
  }
  parse(data) {
    const end = data.lastIndexOf(10) + 1;
    const tail = end !== data.length;
    const events = [];
    if (end) {
      let decoded; try { decoded = new TextDecoder('utf-8',{fatal:true}).decode(data.subarray(0,end)); } catch { throw malformed(this.path,1,'invalid UTF-8'); }
      decoded.slice(0,-1).split('\n').forEach((line,index) => {
        let event; try { event = JSON.parse(line); } catch { throw malformed(this.path,index+1); }
        validateEvent(event,this.path,index+1); events.push(event);
      });
    }
    return {events, tail};
  }
  readEvents() { return this.parse(this.bytes()).events; }
  snapshot() { const data = this.bytes(); return {events:this.parse(data).events, digest:hash(data)}; }
  lockMetadata() {
    try {
      privateDirectory(this.lockPath);
      const fd = regular(path.join(this.lockPath,'owner.json'),C.O_RDONLY,{missing:true});
      if (fd === null) return null;
      try { const value = JSON.parse(fs.readFileSync(fd,'utf8')); return isObject(value) ? value : null; } finally { fs.closeSync(fd); }
    } catch (e) { if (e.code === 'ENOENT' || e instanceof SyntaxError) return null; throw e; }
  }
  removeLock(token) {
    this.check();
    try {
      const before = fs.lstatSync(this.lockPath);
      if (this.lockMetadata()?.token !== token || !same(before,fs.lstatSync(this.lockPath))) return;
      fs.unlinkSync(path.join(this.lockPath,'owner.json')); fs.rmdirSync(this.lockPath);
    } catch (error) {
      if (!['ENOENT','ENOTEMPTY','EEXIST'].includes(error.code)) throw error;
    }
  }
  acquireLock() {
    const token = crypto.randomUUID(); const deadline = performance.now()+5000;
    while (true) {
      this.check(true);
      try { fs.mkdirSync(this.lockPath,{mode:0o700}); break; }
      catch (e) {
        if (e.code !== 'EEXIST') throw e;
        const metadata = this.lockMetadata();
        if (metadata?.host === os.hostname() && nonempty(metadata.token) && Number.isSafeInteger(metadata.pid) && metadata.pid > 0 && Date.now()-Date.parse(metadata.created_at)>300000) {
          let absent = false; try { process.kill(metadata.pid,0); } catch (error) { absent = error.code === 'ESRCH'; }
          if (absent) { this.removeLock(metadata.token); continue; }
        }
        if (performance.now() >= deadline) throw new PapercutsError('lock_timeout',`Timed out waiting for journal lock: ${this.lockPath}`,{retryable:true});
        Atomics.wait(new Int32Array(new SharedArrayBuffer(4)),0,0,25+Math.random()*75);
      }
    }
    try { writeFile(path.join(this.lockPath,'owner.json'),stableJSON({token,pid:process.pid,host:os.hostname(),created_at:new Date().toISOString()})+'\n'); }
    catch (e) { try { fs.rmdirSync(this.lockPath); } catch {} throw e; }
    return token;
  }
  mutation(callback, {allowTail = false} = {}) {
    this.check(true); this.parentIdentity = fs.lstatSync(this.directory);
    let token;
    try {
      token = this.acquireLock(); const data = this.bytes(); const {events,tail} = this.parse(data);
      if (tail && !allowTail) throw new PapercutsError('malformed_journal',`Journal has an incomplete final record: ${this.path}`,{suggested_fix:'Run doctor --repair-tail before mutating the journal.'});
      foldEvents(events,this.path);
      return callback(events,data,tail);
    } finally { try { if (token) this.removeLock(token); } finally { this.parentIdentity = null; } }
  }
  appendEventsLocked(events) {
    if (!events.length) return;
    const data = serializeEvents(events); this.check(true);
    const fd = regular(this.path,C.O_WRONLY|C.O_APPEND,{create:true});
    try { fs.writeFileSync(fd,data); fs.fsyncSync(fd); } finally { fs.closeSync(fd); }
    syncDirectory(this.directory);
  }
  doctor({repair_tail = false} = {}) {
    const inspect = (events,data,tail) => {
      foldEvents(events,this.path); let repaired = false;
      if (tail && repair_tail) {
        this.check(); const fd = regular(this.path,C.O_WRONLY);
        try { fs.ftruncateSync(fd,data.lastIndexOf(10)+1); fs.fsyncSync(fd); } finally { fs.closeSync(fd); }
        data = data.subarray(0,data.lastIndexOf(10)+1); tail = false; repaired = true;
      }
      return {healthy:!tail,repaired,incomplete_tail:tail,byte_count:data.length,event_count:events.length};
    };
    if (repair_tail) return this.mutation(inspect,{allowTail:true});
    const data = this.bytes(), {events,tail} = this.parse(data); return inspect(events,data,tail);
  }
  replaceLocked(events,timestamp) {
    const replacement = serializeEvents(events), original = this.bytes();
    const backups = path.join(this.directory,'papercuts.backups');
    this.check(); privateDirectory(backups,true);
    const backup = path.join(backups,`papercuts-${timestamp.toISOString().replace(/[-:]/g,'')}-${crypto.randomUUID()}.jsonl`);
    writeFile(backup,original); syncDirectory(backups);
    const temporary = path.join(this.directory,`.papercuts-${crypto.randomUUID()}.tmp`);
    try {
      this.check(); writeFile(temporary,replacement); this.check();
      // Refuse a journal replaced by a link while the lock was held.
      const fd = regular(this.path,C.O_RDONLY,{missing:true}); if (fd !== null) fs.closeSync(fd);
      fs.renameSync(temporary,this.path); syncDirectory(this.directory);
    } finally { try { fs.unlinkSync(temporary); } catch (e) { if (e.code !== 'ENOENT') throw e; } }
    return {backup,before_bytes:original.length,after_bytes:replacement.length};
  }
}
